### DISCUSSION TITLE
[Hackathon usecase] Make Morphir's Engine for compiling Elm to Morphir and interpreting Morphir IRs available via WebAssembly and WASI.

### DISCUSSION THREAD

**Discussion by DamianReeves** [finos/finos-pmcs#262] [Ideas]
# Business Problem

Morphir offer's businesses a way to represent their business logic using a data format (the Morphir IR) that is language agnostic. Despite the format being language agnostic, the Morphir tooling is currently very tied to the Elm programming language, NodeJS, and to a degree the JVM and Scala/Java.

WebAssembly as a platform offers execution flexibility and engines like `Wasmtime` and `Wasmer` run on a variety of platforms. If we took the core Morphir CLI engine and exposed it through a WebAssembly interface, we would enable other programming languages and ecosystems to integrate easier with the toolchain and to ultimately leverage Morphir's interpreter and build pipeline.

# Expected Outcome

While full integration of the Morphir toolchain into WebAssembly will most likely take more time than we have for this hackathon, we can do exploration and initial work to enable 2 usecases:

1. Creating an instance of the CLI engine which is exposed via a WASM/WASI interface. Initial focus for the CLI is on providing the `make` functionality.
2. Creating a wrapper around Morphir's interpreter and exposing it through WASM.

Both of these initiatives will benefit Morphir and its users by making integration with Morphir more accessible and not limited solely to executing the CLI and using the generated code.

**Comment by TheJuanAndOnly99** [comment]
Closing discussion in preparation for [this years hackathon](https://www.finos.org/hosted-events/2024-10-25-citi-india-hackathon-supported-by-finos). Feel free to reopen and update if its still relevant.
